### ISSUE TITLE
Could not connect to Exact Error 401: {"error":"unauthorized_client","error_description":"Old refresh token used."}

### ISSUE THREAD

**Issue by fregoe** [picqer/exact-php-client#644]
Sometimes, without any reason, i receive this error:
Could not connect to Exact Error 401: {"error":"unauthorized_client","error_description":"Old refresh token used."}

Then I need to manually clear the access, refresh en authorization code and request all new.

We use this packages for automatic sync so this is an issue because requesting new authorization codes requires browser interaction.

**Comment by remkobrenters** [comment]
This error is the result of an implementation error you made somewhere. Due to the way OAuth works, the only solution after encountering this error is to ask the user to give new consent so you get a new set of tokens that are valid again.

Please see https://github.com/picqer/exact-php-client/issues/640 (or some of the numerous other issues reported by people with refresh token issues).

**Comment by remkobrenters** [comment]
The reason I am a bit strong about the wrong implementation is that we use this library (without any extra logic for this part) in hundreds of connections of which some of them are heavy users (premium users) with thousands of requests. 

The problem (and the provided solution - thanks for chipping in) as I see it come down to concurrent requests to the same connection resulting in the token refreshes of the requests breaking each-other. This can be prevented by locking the tokens (like you suggest) or by making sure that requests for the same connection are queued so they are not performed at the same time.

The thing that I wonder: Are you facing this issue with user-facing logic where users perform actions that can be at the same time? As all our implementations are server-to-server and run in background processes where we have a lot of control over the queue and preventing overlapping requests.

**Comment by remkobrenters** [comment]
That would be awesome as it comes by every now and than. It might be caused by something that we do not use in our requests making me so sure that the implementation in base works well.

**Comment by meijdenmedia** [comment]
Yeah, but that means you've just lost the refresh token and you need to manually reconnect, right? How should this be handled in cronjobs f.e.? 🤔
